### PR TITLE
Add data-bundle attributes on javascript and styles resources

### PR DESCRIPTION
**Pull request by aormazabal**
Is useful to have these attributes in resources to filter them through Diazo rules.
Sometimes you need to remove or show this resources, for styling motivations.

It could be possible with the last change in webresource library. Webresource v.1.2

Added attribute on Resource declaration

            PloneStyleResource(
               ...,
                **{'data-bundle': name}
            )

### CODE REVIEW

**Comment by aormazabal** [comment]
> @jenkins-plone-org please run jobs

@jenkins-plone-org please run jobs

**Comment by aormazabal** [comment]
> > @jenkins-plone-org please run jobs
> 
> @jenkins-plone-org please run jobs

@plone/developers I can't pass the jobs because there is a constraint present in https://dist.plone.org/release/6.0.0/constraints.txt
webresource==1.1
It must be webresource==1.2
How can I manage this problem?

**Comment by gotcha** [comment]
I do not know how to answer your question.

Nevertheless, I think you should add a test to support your feature...

**Comment by petschki** [comment]
First you should create a PR on https://github.com/plone/buildout.coredev.git with the updated versions.cfg file https://github.com/plone/buildout.coredev/blob/6.0/versions.cfg#L229 ... you can then run this PR and the coredev PR at once by providing both urls in the jenkins parameters. But as @gotcha said, adding a test which tests this feature would be highly recommended.

**Review by Rudd-O — APPROVED** [review]

**Review by jensens — CHANGES_REQUESTED** [review]
- needs tests
- first bump version in coredev

**Comment by aormazabal** [comment]
> * needs tests
> 
>     * first bump version in coredev

I'm trying to do tests in local for this product but not works.

`I execute something like ./bin/test -s Products.CMFPlone -t test_scripts_viewlet`

But the response is: 

`Total: 0 tests, 0 failures, 0 errors, 0 skipped in 0.000 seconds.`

Some help?

**Comment by aormazabal** [comment]
> > @jenkins-plone-org please run jobs
> 
> @jenkins-plone-org please run jobs

@jenkins-plone-org please run jobs

**Comment by aormazabal** [comment]
> 1. make sure you have added `Products.CMFPlone[test]` in your `[test]` buildout part
> 
>     2. check `bin/test --list-tests` for available tests and check if the `-t <testname>` matches one from the list

Thanks!

**Comment by aormazabal** [comment]
> > 1. make sure you have added `Products.CMFPlone[test]` in your `[test]` buildout part
> >    
> >    2. check `bin/test --list-tests` for available tests and check if the `-t <testname>` matches one from the list
> 
> Thanks!

I don't understand why failed test. A priori these error is nothing related with my tests.
I test in my local **test_collection_deserializer_validates_value** and not fails

`Failure in test test_collection_deserializer_validates_value (plone.restapi.tests.test_dxfield_deserializer.TestDXFieldDeserializer)
Traceback (most recent call last):
  File "/srv/python3.8/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/srv/python3.8/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/srv/python3.8/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/jenkins/workspace/pull-request-6.0-3.8/src/plone.restapi/src/plone/restapi/tests/test_dxfield_deserializer.py", line 490, in test_collection_deserializer_validates_value
    self.assertEqual((b"2", (int,), ""), cm.exception.args)
  File "/srv/python3.8/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/srv/python3.8/lib/python3.8/unittest/case.py", line 1129, in assertTupleEqual
    self.assertSequenceEqual(tuple1, tuple2, msg, seq_type=tuple)
  File "/srv/python3.8/lib/python3.8/unittest/case.py", line 1100, in assertSequenceEqual
    self.fail(msg)
  File "/srv/python3.8/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: Tuples differ: (b'2', (<class 'int'>,), '') != (b'2', <class 'int'>, '')

First differing element 1:
(<class 'int'>,)
<class 'int'>

- (b'2', (<class 'int'>,), '')
?        -              --

+ (b'2', <class 'int'>, '')`

**Comment by aormazabal** [comment]
@plone/developers

I don't understand why failed jenkins test. A priori these error is nothing related with my tests.

Failure in test test_collection_deserializer_validates_value (plone.restapi.tests.test_dxfield_deserializer.TestDXFieldDeserializer)
--> AssertionError: Tuples differ: (b'2', (<class 'int'>,), '') != (b'2', <class 'int'>, '')


I test in my local **test_collection_deserializer_validates_value** and  fails.
In fact, I return the Products.CMFPlone to the master branch and also fails. so these issue is not related to the changes branch [webresource_databundle_attr](https://github.com/plone/Products.CMFPlone/tree/webresource_databundle_attr) that I try to test

Any help?

**Comment by petschki** [comment]
Hi ... I also got this failure in another PR https://github.com/plone/plone.staticresources/pull/261 ... seems unrelated. I try to figure out why right now.

File: Products/CMFPlone/resources/browser/resource.py [hunk 1]
```diff
@@ -187,6 +187,7 @@ def check_dependencies(bundle_name, depends, bundles):
                     async_=record.load_async or None,
                     defer=record.load_defer or None,
                     integrity=not external,
+                    **{'data-bundle': name},
```
Comment (mauritsvanrees) [inline Products/CMFPlone/resources/browser/resource.py:190]: The following looks simpler and should be the same, right? Or am I missing something?

```suggestion
                    data-bundle=name,
```

Same a few lines further.

Comment (mauritsvanrees) [reply Products/CMFPlone/resources/browser/resource.py:190]: Ah, I see, `data-bundle` is probably illegal as parameter and you would get a SyntaxError. Never mind.

Comment (aormazabal) [reply Products/CMFPlone/resources/browser/resource.py:190]: Exactly, In his way, I contributed to a new feature in webresource package v1.2 to solve this inconvenient. Now accepts more parameters and could be pass with a dict of attributes.

https://github.com/conestack/webresource/blob/master/webresource/_api.py#L134

**Comment by petschki** [comment]
It seems, that this test buildout gets `zope.schema>=7.0.0` which changed the `_type` in `schema.Int` from `(int, )`  to `int` and therefore the traceback changed when a validation error is raised ... see https://github.com/zopefoundation/zope.schema/blob/master/src/zope/schema/_bootstrapfields.py#L925 ... 

@mauritsvanrees @gforcada it would be helpful if Jenkins makes a `buildout annotate` to determine the picked versions of any package...

**Comment by petschki** [comment]
So pinning `zope.schema = 6.2.1` should solve the error on Jenkins

**Comment by aormazabal** [comment]
> zope.schema = 6.2.1

So, can I change the [version.cfg in corede](https://github.com/plone/buildout.coredev/blob/6.0/versions.cfg) adding zope.schema = 6.2.1?

**Comment by petschki** [comment]
I think we should switch the comments here on line 8 and 7 back to Zope 5.7.x since we're obviously not ready for Zope 5.8 yet https://github.com/plone/buildout.coredev/blob/6.0/versions.cfg#L8 ... but I'd like some other oppinions on that.

/cc @mauritsvanrees

**Comment by mauritsvanrees** [comment]
Zope 5.8 will be fine, but the zope.schema update in the Zope master versions breaks it. And some more changes to its versions are in the works, so I will change coredev to use the 5.8 release and its official versions. I am pushing a change now.

BTW, I don't know why the latests Jenkins tests pass, and don't fail in the same way as the PR tests.

**Review by mauritsvanrees — APPROVED** [review]
LGTM, thanks!
I expect the tests will be green now, let's see:

@jenkins-plone-org please run jobs

**Comment by aormazabal** [comment]
> LGTM, thanks! I expect the tests will be green now, let's see:
> 
> @jenkins-plone-org please run jobs

Unfortunately it does not seem . It still returns the same error in:
test_collection_deserializer_validates_value (plone.restapi.tests.test_dxfield_deserializer.TestDXFieldDeserializer)

**Comment by mauritsvanrees** [comment]
> > LGTM, thanks! I expect the tests will be green now, let's see:
> > @jenkins-plone-org please run jobs
> 
> Unfortunately it does not seem . It still returns the same error in: test_collection_deserializer_validates_value (plone.restapi.tests.test_dxfield_deserializer.TestDXFieldDeserializer)

The tests are still running on Jenkins. You must be looking at an older run.

**Comment by jensens** [comment]
@aormazabal do **never** merge own PRs - this is upon other core developers.

**Comment by mauritsvanrees** [comment]
Indeed, that is the rule.
There *were* two approving reviews though, and the tests were green, so don't feel bad about it. Thank you for this improvement!

Note that this rule can differ per community. In the [Zope Foundation](https://github.com/zopefoundation/) repositories the rule is the other way around: you merge your own PRs, when tests are green, there is at least one approving review, and there are no unresolved comments.

**Comment by stevepiercy** [comment]
For this repo, if there is a rule that you want to enforce and avoid self-merging by non-core members, then set it under Settings > Branches > Branch protection rules.

https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/managing-a-branch-protection-rule

> Note that this rule can differ per community.

Merging rules also differ within the Plone community.

- https://github.com/plone/documentation - At least one approval, all builds pass, primary branch is protected to allow only members of the Documentation Team or "Organization and repository administrators" to merge, OK to merge your own PR. There are rare exceptions, such as routine update of git submodule tips, or emergency fixes.
- https://github.com/plone/training - No rules. It's the wild west! pew! pew!
- https://github.com/plone/volto - At least one approval, all builds pass, OK to merge your own PR.

The bottom line is to set branch protection rules according to your stated policy.

**Comment by aormazabal** [comment]
> Indeed, that is the rule. There _were_ two approving reviews though, and the tests were green, so don't feel bad about it. Thank you for this improvement!
> 
> Note that this rule can differ per community. In the [Zope Foundation](https://github.com/zopefoundation/) repositories the rule is the other way around: you merge your own PRs, when tests are green, there is at least one approving review, and there are no unresolved comments.

Sorry to everyone!!

I missunderstood your comment...

https://github.com/plone/Products.CMFPlone/issues/3707#issuecomment-1378976165
```
This is a minor improvement and does not need a PLIP in my opinion.
You have made a PR for this: https://github.com/plone/Products.CMFPlone/pull/3708
Should be green soon now, and then can be merged.
```

**Comment by mauritsvanrees** [comment]
Sorry, I was not clear in that comment on who should be doing the merging officially.

**Comment by aormazabal** [comment]
> Sorry, I was not clear in that comment on who should be doing the merging officially.

Sorry to you!! Is my fault, there is a lot information about how to proceed. And thanks for help me!!!!!

**Comment by jensens** [comment]
Sorry, I hope it did not sound that harsh. I agree with @stevepiercy to use the features from Github to enforce such rules instead of blaming our valuable community members. Thanks for your contribution!
